File: OsmAndWear/src/net/osmand/plus/settings/backend/preferences/PreferenceWithListener.java
package net.osmand.plus.settings.backend.preferences;

import net.osmand.StateChangedListener;

import java.lang.ref.WeakReference;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

public abstract class PreferenceWithListener<T> implements OsmandPreference<T> {
	private List<WeakReference<StateChangedListener<T>>> l;

	@Override
	public synchronized void addListener(StateChangedListener<T> listener) {
		if (l == null) {
			l = new LinkedList<WeakReference<StateChangedListener<T>>>();
		}
		if (!l.contains(new WeakReference<StateChangedListener<T>>(listener))) {
			l.add(new WeakReference<StateChangedListener<T>>(listener));
		}
	}

	public synchronized void fireEvent(T value) {
		if (l != null) {
			Iterator<WeakReference<StateChangedListener<T>>> it = l.iterator();
			while (it.hasNext()) {
				StateChangedListener<T> t = it.next().get();
				if (t == null) {
					it.remove();
				} else {
					t.stateChanged(value);
				}
			}
		}
	}

	@Override
	public synchronized void removeListener(StateChangedListener<T> listener) {
		if (l != null) {
			Iterator<WeakReference<StateChangedListener<T>>> it = l.iterator();
			while (it.hasNext()) {
				StateChangedListener<T> t = it.next().get();
				if (t == listener) {
					it.remove();
				}
			}
		}
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/settings/backend/preferences/PreferenceWithListener.kt
package net.osmand.shared.settings.backend.preferences

import co.touchlab.stately.concurrency.Synchronizable
import co.touchlab.stately.concurrency.synchronize
import net.osmand.shared.api.KStateChangedListener

abstract class PreferenceWithListener<T> : OsmandPreference<T> {

	private var syncObj = Synchronizable()

	private val listeners = mutableListOf<KStateChangedListener<T>?>()

	override fun addListener(listener: KStateChangedListener<T>) {
		syncObj.synchronize {
			if (listeners.none { it === listener }) {
				listeners.add(listener)
			}
		}
	}

	fun fireEvent(value: T) {
		syncObj.synchronize {
			val iterator = listeners.iterator()
			while (iterator.hasNext()) {
				val listener = iterator.next()
				if (listener == null) {
					iterator.remove()
				} else {
					listener.stateChanged(value)
				}
			}
		}
	}

	override fun removeListener(listener: KStateChangedListener<T>) {
		syncObj.synchronize {
			val iterator = listeners.iterator()
			while (iterator.hasNext()) {
				val registeredListener = iterator.next()
				if (registeredListener === listener) {
					iterator.remove()
				}
			}
		}
	}
}